cpp/src/arrow/ipc/reader.cc

```diff
@@ -107,6 +108,23 @@ Status InvalidMessageType(MessageType expected, MessageType actual) {
 // ----------------------------------------------------------------------
 // Record batch read path
 
+/// \brief Structure to keep common arguments to be passed
+struct IpcReadContext {
+  IpcReadContext(DictionaryMemo* memo, const IpcReadOptions& option, DictionaryKind* kind,
+                 bool swap)
+      : dictionary_memo(memo), options(option), kind(kind), swap_endian(swap) {}
+
+  DictionaryMemo* dictionary_memo;
+
+  const IpcReadOptions& options;
+
+  DictionaryKind* kind;
```

Comment:
Having a mutable pointer here for an enum seems odd, can you explain?

Reply:
This style, which uses a mutable pointer, follows the original code. Here is an example. My guess is that this code needs to have a special state `uninitialized`.

https://github.com/apache/arrow/blob/master/cpp/src/arrow/ipc/reader.cc#L716-L726

Reply:
Based on [this suggestion](https://github.com/apache/arrow/pull/7507#discussion_r513782671), I dropped `kind` from `IpcReadContext`.